File: packages/api-models-ts/src/RecurringApis.ts
/**
 * Request and response models for `/api/recurring_api`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Schemas from "./Schemas.ts";
import * as Users from "./Users.ts";

/**
 * A recurring API call scheduled by a user (upstream `RecurringAPI`).
 *
 * `owner` is always loaded (`lazy="selectin"` upstream). `payload` is
 * free-form JSON, and the single-object endpoint returns it exactly as stored,
 * which may still be a JSON string.
 *
 * @since 1.0.0
 * @category Models
 */
export const RecurringApi = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        endpoint: Schemas.NullishString,
        method: Schemas.NullishString,
        payload: Schemas.nullish(v.union([Schemas.JsonObject, v.string()])),
        next_call: Schemas.NullishTimestamp,
        call_delay: Schemas.NullishNumber,
        number_of_retries: Schemas.NullishInteger,
        active: Schemas.NullishBoolean,
        owner_id: Schemas.NullishInteger,
        owner: Schemas.nullish(Users.User),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type RecurringApi = v.InferOutput<typeof RecurringApi>;

/**
 * Payload for scheduling a recurring API call.
 *
 * `method` is upper-cased by the server and must end up as `"GET"` or
 * `"POST"`, `payload` must be a valid JSON string, `next_call` is any
 * arrow-parseable timestamp, `call_delay` is in days, and `number_of_retries`
 * may not exceed `10`.
 *
 * @since 1.0.0
 * @category Models
 */
export interface RecurringApiPost {
    readonly endpoint: string;
    readonly method: string;
    readonly next_call: string;
    readonly call_delay: number;
    readonly payload: string;
    readonly number_of_retries?: number | undefined;
}

/**
 * Result of scheduling a recurring API call.
 *
 * @since 1.0.0
 * @category Models
 */
export const RecurringApiPostResponse = Schemas.model(v.strictObject({ id: Schemas.Integer }));

/**
 * @since 1.0.0
 * @category Models
 */
export type RecurringApiPostResponse = v.InferOutput<typeof RecurringApiPostResponse>;

File: packages/api-models-py/skyportal_py_models/recurring_apis.py
"""Request and response models for ``/api/recurring_api``."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict

from skyportal_py_models.users import User


class RecurringAPI(BaseModel):
    """A recurring API call scheduled by a user (upstream ``RecurringAPI``).

    ``owner`` is always loaded (``lazy="selectin"`` upstream). ``payload``
    is free-form JSON, and the single-object endpoint returns it exactly as
    stored, which may still be a JSON string.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime | None = None
    modified: datetime | None = None
    endpoint: str | None = None
    method: str | None = None
    payload: dict[str, Any] | str | None = None
    next_call: datetime | None = None
    call_delay: float | None = None
    number_of_retries: int | None = None
    active: bool | None = None
    owner_id: int | None = None
    owner: User | None = None


class RecurringAPIPost(BaseModel):
    """Payload for scheduling a recurring API call."""

    model_config = ConfigDict(extra="forbid")

    endpoint: str
    method: str
    next_call: str
    call_delay: float
    payload: str
    number_of_retries: int | None = None


class RecurringAPIPostResponse(BaseModel):
    """Result of scheduling a recurring API call."""

    model_config = ConfigDict(extra="forbid")

    id: int
